src/plugins/mer/merhardwaredevicewizardpages.cpp

```diff
@@ -237,6 +263,118 @@ QString MerHardwareDeviceWizardSelectionPage::detectDeviceName(
+    }
+
+    if (QVersionNumber::fromString(osVersion) < QVersionNumber(3, 2, 1)) {
+        *ok = true;
+        qDebug() << "Sailfish OS" << osVersion << "assume that packages are installed";
+        return true;
+    }
+
+    runner.run("rpm -q patterns-sailfish-sdk-client-tools", sshParams);
+    loop.exec();
+
+    if (!runner.lastConnectionErrorString().isEmpty()
+            || runner.processExitStatus() != SshRemoteProcess::NormalExit) {
+        *ok = false;
+        qWarning() << "Failed to query install status of patterns-sailfish-sdk-client-tools";
+        return false;
+    }
+
+    if (runner.processExitCode() != 0) {
+        *ok = true;
+        qInfo() << "patterns-sailfish-sdk-client-tools is not installed";
+        return false;
+    }
+
+    qInfo() << "patterns-sailfish-sdk-client-tools is installed";
+    *ok = true;
+    return true;
+}
+
+bool MerHardwareDeviceWizardSelectionPage::askInstallSdkClientTools(QString &password)
+{
+    QInputDialog passwordDialog;
+    passwordDialog.setWindowTitle(tr("SDK tools needed"));
+    passwordDialog.setLabelText(tr("Enter developer mode password to install SDK tools."));
+    passwordDialog.setInputMode(QInputDialog::TextInput);
+    passwordDialog.setTextEchoMode(QLineEdit::Password);
+    passwordDialog.setCancelButtonText(tr("Abort"));
+    passwordDialog.setOkButtonText(tr("Install"));
+    passwordDialog.exec();
+    password = passwordDialog.textValue();
+    return passwordDialog.result() == QInputDialog::Accepted;
+}
+
+bool MerHardwareDeviceWizardSelectionPage::installSdkClientTools(
+        const SshConnectionParameters &sshParams, QString &password)
+{
+    SshRemoteProcessRunner runner;
+    QEventLoop loop;
+    QByteArray passwordData(password.toLatin1());
+    passwordData.append("\n");
+    connect(&runner, &SshRemoteProcessRunner::connectionError,
+            &loop, &QEventLoop::quit);
+    connect(&runner, &SshRemoteProcessRunner::processClosed,
+            &loop, &QEventLoop::quit);
+    connect(&runner, &SshRemoteProcessRunner::processStarted,
+            [&runner, &passwordData] { runner.writeDataToProcess(passwordData); });
+    runner.run("devel-su pkcon --plain --noninteractive refresh", sshParams);
+    loop.exec();
+
+    if (!runner.lastConnectionErrorString().isEmpty()
+            || runner.processExitStatus() != SshRemoteProcess::NormalExit
+            || runner.processExitCode() != 0) {
+        qWarning() << "Failed to refresh packages";
+        return false;
+    }
+
+    runner.run("devel-su pkcon --plain --noninteractive install "
+               "patterns-sailfish-sdk-client-tools", sshParams);
+    loop.exec();
+
+    if (!runner.lastConnectionErrorString().isEmpty()
+            || runner.processExitStatus() != SshRemoteProcess::NormalExit
+            || runner.processExitCode() != 0) {
+        qWarning() << "Failed to install patterns-sailfish-sdk-client-tools";
+        return false;
+    }
+
+    passwordData.truncate(0);
```

Comment:
Why only in the successful case? Also if I am correct that you do it to clear the password from process memory, I wouldn't be that optimistic about the effect of truncate().

Reply:
No reason, just a simple mistake. Also meant that truncate(0) to be clear() which is used in devicelock for this purpose.